[feature] File view page: link to remote location

caseyrollins opened the pull request:
#### Purpose
- Adds a link to the file view page which links to the provider's external view of the file.
- Adds a "Go to ..." button to the file tree which also links to the provider's external view. 
- Both links utilize WaterButler's ```WebViewHandler``` to create/retrieve the external view link.
- Addresses [#2926](https://github.com/CenterForOpenScience/osf.io/issues/2926)
- Addons Supported:
 - [x] GitHub
 - [x] Google Drive
 - [x] Figshare
 - [x] Dropbox
 - [x] Box
- Screenshot:
 - File detail page:
![screen shot 2015-06-25 at 1 36 01 pm](https://cloud.githubusercontent.com/assets/7913604/8361128/3ef5e89c-1b3f-11e5-9900-29445c22885d.png)

 - File tree:
![screen shot 2015-07-08 at 1 47 09 pm](https://cloud.githubusercontent.com/assets/7913604/8604223/3ea844e0-264b-11e5-8dc0-802ede619603.png)

#### Notes
- Dependent on WaterButler PR [#52](https://github.com/CenterForOpenScience/waterbutler/pull/52)
- Dependent on WaterButler PR [#56](https://github.com/CenterForOpenScience/waterbutler/pull/56)

lyndsysimon commented on website/addons/base/views.py:
```diff
@@ -435,6 +435,7 @@ def addon_view_file(auth, node, node_addon, guid_file, extras):
         # Note: must be called after get_or_start_render. This is really only for github
         'size': size,
         'extra': json.dumps(getattr(guid_file, 'extra', {})),
+        'source_url': getattr(guid_file, 'source_url', None),
```
Generally, I'd only use `getattr()` if the attribute I was attempting to access wasn't known ahead of time. 

Have you considered setting `GuidFile.source_url` at the level on the parent class, and defaulting it to `None`? That would eliminate the need for this logic and ensure that if this property is needed by a sibling class that it doesn't get implemented a second time.

lyndsysimon replied:
After seeing that you've defined it for several subclasses of `GuidFile`, please go ahead and make it a property of the parent class and remove this logic.

caseyrollins commented:
@lyndsysimon PR is ready for review! :octopus:

lyndsysimon commented on website/addons/base/__init__.py:
```diff
@@ -208,6 +208,7 @@ class GuidFile(GuidStoredObject):
     _metadata_cache = None
     _id = fields.StringField(primary=True)
     node = fields.ForeignField('node', required=True, index=True)
+    source_url = None
```
This isn't necessary, as source_url is defined as a property method, below.

lyndsysimon commented on website/addons/base/__init__.py:
```diff
@@ -345,6 +346,11 @@ def deep_url(self):
 
     @property
+    def source_url(self):
+        extra = self._metadata_cache['extra']
```
Will there be any case where `self._metadata_cache` doesn't contain the key `extra`?

If so, perhaps this would be clearer:

```python
    @property
    def source_url(self):
        try:
            return self._metadata_cache['extra']['source_url']
        except KeyError:
            return None
```

If you're confident that `extra` will always be there, make no changes :)

caseyrollins replied:
```extra``` will always exist because it is a property of WaterButler's [BaseMetadata](https://github.com/CenterForOpenScience/waterbutler/blob/develop/waterbutler/core/metadata.py#L13-27) class :+1:

lyndsysimon commented:
First pass complete, pending response to comments.

caseyrollins commented:
@lyndsysimon First pass changes done :cake:

chrisseto commented on website/static/js/filepage/index.js:
```diff
@@ -198,12 +199,17 @@ var FileViewPage = {
         m.render(document.getElementById('toggleBar'), m('.btn-toolbar[style=margin-top:20px]', [
-            ctrl.canEdit() ? m('.btn-group', {style: 'margin-left: 0;'}, [
-                m('.btn.btn-sm.btn-danger.file-delete', {onclick: $(document).trigger.bind($(document), 'fileviewpage:delete')}, 'Delete')
+            ctrl.sourceUrl !== 'None' ? m('.btn-group', {}, [
```
`None` would be serialized as `null` in javascript not the string value `'None'`

caseyrollins replied:
Wouldn't a ```None``` source_url be set to the string ```'None'``` [here](https://github.com/caseyrollins/osf.io/blob/feature/source-urls/website/templates/project/view_file.mako#L152) first though?

lyndsysimon replied:
I think you're right.

Also: ewwwwwwwwww. :disappointed:

chrisseto commented on website/templates/project/view_file.mako:
```diff
@@ -149,6 +149,7 @@
         file: {
             size: ${size},
             extra: ${extra},
+            sourceUrl: '${source_url}',
```
You need to use `js_str` here
You may also want to add an if in so that it actually gets set to `null` or `undefined` as per the above comment. but to quote sloria, follow your :heart:

lyndsysimon commented on website/templates/project/view_file.mako:
```diff
@@ -149,6 +149,7 @@
         file: {
             size: ${size},
             extra: ${extra},
+            sourceUrl: '${source_url | js_str}',
```
How about this:

```
sourceUrl: '${(source_url or '') | js_str}',
```

That way, at least you can check for the value to be an empty string if it isn't provided, instead of the string `"None"`. The string representation of the `None` object feeels dirty to me.

I'll leave it up to you, though. Let me know either way and I'll pass it on to Steve.

caseyrollins replied:
I'm good with this. Making the change now.

lyndsysimon commented:
Good deal. Ready for @sloria. :yum:

chrisseto commented:
My final thought is how do we feel about calling it "source url"? It seems a tad misleading as its for viewing a file via the original content provider.

caseyrollins commented:
I agree with @chrisseto's point about the naming. Would something like "external view link" make more sense?

chrisseto commented:
@caseyrollins Lets go with "view url"

caseyrollins commented:
@chrisseto :+1:

sloria commented on website/addons/base/views.py:
```diff
@@ -435,6 +435,7 @@ def addon_view_file(auth, node, node_addon, guid_file, extras):
         # Note: must be called after get_or_start_render. This is really only for github
         'size': size,
         'extra': json.dumps(getattr(guid_file, 'extra', {})),
+        'view_url': guid_file.view_url,
```
This should probably go in the `urls` dict above. Also, `view_url` is ambiguous (could be confused with the URL to view the file on OSF)--how about `external`?

sloria commented on website/static/js/filepage/index.js:
```diff
@@ -198,12 +199,17 @@ var FileViewPage = {
-            ctrl.canEdit() ? m('.btn-group', {style: 'margin-left: 0;'}, [
-                m('.btn.btn-sm.btn-danger.file-delete', {onclick: $(document).trigger.bind($(document), 'fileviewpage:delete')}, 'Delete')
+            ctrl.viewUrl ? m('.btn-group', {}, [
+                m('a', {href: ctrl.viewUrl, target: '_blank', class: 'btn btn-sm btn-success'}, 'View ', [
```
We already have a "View" button in the top-right corner, so this text should probably be more unique. Maybe "View on <addon>", e.g. "View on Dropbox"?

sloria replied:
@GaryKriebel @lbanner Spoke with @caseyrollins , and she has considered the alternative I suggested above, but changing it to "View on <addon>" can make the button awkwardly large. Perhaps you or a Community member could provide input?

GaryKriebel replied:
I'll pull the code and give it a look.

GaryKriebel replied:
I currently see this:
![screen shot 2015-06-24 at 2 04 10 pm](https://cloud.githubusercontent.com/assets/9437946/8337568/f4abcdbc-1a79-11e5-8885-a50d9c84930f.png)

sloria replied:
@GaryKriebel The problem is that there are now 2 "View" buttons that do different things.

I suggested changing to "View on Dropbox" but that makes the button large (which may or may not be a problem).

@caseyrollins Also, I just checked other parts of the site, and it appears that the icon should go *before* the text in the button.

sloria replied:
The other suggestion I have is to change the blue "View" text to "Content", but this also is less than ideal.

sloria replied:
Another option: Instead of a button, show the provider's logo as the external link.

sloria commented:
Looks good. :v: Just a few minor naming issues and this should be good to merge.

@chrisseto Is the supporting waterbutler PR merged?

chrisseto commented:
@sloria Doing one more pass and will merge it

caseyrollins commented:
@sloria Made those code review changes, removed the "View" button as discussed with @GaryKriebel, and updated the PR description to reflect that change. :cake:

sloria commented on website/templates/project/view_file.mako:
```diff
@@ -51,6 +51,7 @@
   <div id="fileViewPanelLeft" class="col-sm-9 panel-expand">
     <div class="row">
       <div id="mfrIframeParent" class="col-sm-9">
+        <p class="text-muted">View this file on <a href=${urls['external']} target="_blank">${provider}</a>.</p>
```
`provider` is not human-friendly

![osf___hh_nlp_talk_draft_pdf](https://cloud.githubusercontent.com/assets/2379650/8463559/53ebafe2-2009-11e5-9a48-afe2dfd05efc.png)

We should display the addon's human-friendly representation defined in each addon's config as `FULL_NAME`.

sloria commented:
Just a minor change and this is good to merge.

sloria commented:
Once the "View on ..." button has been added to the Files page toolbar, this should be good to merge.

sloria commented:
Holding off on next pass, pending changes to waterbutler. @caseyrollins Please ping me when ready.

caseyrollins commented:
@sloria, I've added the "Go to ..." button to the files toolbar, and updated the PR description to reflect the new changes (including a dependency on [this](https://github.com/CenterForOpenScience/waterbutler/pull/56) PR). 

Ready for another pass :tropical_fish:

sloria commented:
Thanks. I will look at this today.

sloria commented:
I will do final pass once https://github.com/CenterForOpenScience/waterbutler/pull/56 is merged.

sloria commented:
My popup blocker (on Chrome) isn't allowing me to view any of the "View on" links. I think this is because of the "_blank" attribute.

sloria commented:
The button on the file page should be "View on ...", not "Go to ..."

sloria commented on website/static/js/filepage/index.js:
```diff
@@ -63,6 +65,16 @@ var FileViewPage = {
             return false;
         });
 
+        $(document).on('fileviewpage:externalView', function() {
```
Why is this custom event necessary? Why not just pass this callback as the `onclick` handler?

sloria commented:
When you send the new PR, make sure the button order is consistent across addons.

GitHub:

![osf___westerlies_virus_alignment_elements_aviar_files](https://cloud.githubusercontent.com/assets/2379650/8730213/85b77f78-2bbd-11e5-858c-7b2dd259688e.png)


Google Drive:

![osf___westerlies_virus_alignment_elements_aviar_files](https://cloud.githubusercontent.com/assets/2379650/8730507/169c42c0-2bbf-11e5-8c8e-a92a343924d3.png)
